src/interpreter/plugin/ConditionalAggregationPlugin.ts

```diff
@@ -0,0 +1,331 @@
+/**
+ * @license
+ * Copyright (c) 2022 Handsoncode. All rights reserved.
+ */
+
+import {CellError, ErrorType} from '../../Cell'
+import {ErrorMessage} from '../../error-message'
+import {Maybe} from '../../Maybe'
+import {ProcedureAst} from '../../parser'
+import {Condition, CriterionFunctionCompute} from '../CriterionFunctionCompute'
+import {InterpreterState} from '../InterpreterState'
+import {
+  getRawValue,
+  InternalScalarValue,
+  InterpreterValue,
+  isExtendedNumber,
+  RawInterpreterValue,
+  RawScalarValue
+} from '../InterpreterValue'
+import {SimpleRangeValue} from '../SimpleRangeValue'
+import {ArgumentTypes, FunctionPlugin, FunctionPluginTypecheck} from './FunctionPlugin'
+
+class AverageResult {
+  public static empty = new AverageResult(0, 0)
+
+  constructor(
+    public readonly sum: number,
+    public readonly count: number,
+  ) {}
+
+  public static single(arg: number): AverageResult {
+    return new AverageResult(arg, 1)
+  }
+
+  public compose(other: AverageResult) {
+    return new AverageResult(this.sum + other.sum, this.count + other.count)
+  }
+
+  public averageValue(): Maybe<number> {
+    if (this.count > 0) {
+      return this.sum / this.count
+    } else {
+      return undefined
+    }
+  }
+}
+
+/** Computes key for criterion function cache */
+function conditionalAggregationFunctionCacheKey(functionName: string): (conditions: Condition[]) => string {
+  return (conditions: Condition[]): string => {
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const conditionsStrings = conditions.map((c) => `${c.conditionRange.range!.sheet},${c.conditionRange.range!.start.col},${c.conditionRange.range!.start.row}`)
+    return [functionName, ...conditionsStrings].join(',')
+  }
+}
+
+function zeroForInfinite(value: InternalScalarValue) {
+  if (isExtendedNumber(value) && !Number.isFinite(getRawValue(value))) {
+    return 0
+  } else {
+    return value
+  }
+}
+
+function mapToRawScalarValue(arg: InternalScalarValue): Maybe<CellError | RawScalarValue> {
+  if (arg instanceof CellError) {
+    return arg
+  }
+
+  if (isExtendedNumber(arg)) {
+    return getRawValue(arg)
+  }
+
+  return undefined
+}
+
+export class ConditionalAggregationPlugin extends FunctionPlugin implements FunctionPluginTypecheck<ConditionalAggregationPlugin> {
```

Comment:
I renamed the class `SumifPlugin` to `ConditionalAggregationPlugin`. It may be used as an argument for `registerFunctionPlugin()` and `unregisterFunctionPlugin()` but its never mentioned in our docs so I consider this as non-breaking change. @warpech WDYT?

Reply:
Yeah, I think it is an internal change.